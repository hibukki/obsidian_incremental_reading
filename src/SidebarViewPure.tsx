import React from "react";
import { Rating } from "ts-fsrs";
import { CardStats, IntervalPreviews } from "./types";

export interface SidebarViewPureProps {
	dueCount: number;
	totalCount: number;
	status: string;
	statusHappy: boolean;
	showDifficultyButtons: boolean;
	cardStats: CardStats | null;
	intervalPreviews: IntervalPreviews | null;
	onShowNext: () => void;
	onAddToQueue: () => void;
	onShowQueue: () => void;
	onMarkAgain: () => void;
	onMarkHard: () => void;
	onMarkGood: () => void;
	onMarkEasy: () => void;
}

/**
 * Pure presentational component for the Incremental Reading sidebar.
 * All data comes from props, making it easy to test.
 */
export const SidebarViewPure: React.FC<SidebarViewPureProps> = ({
	dueCount,
	totalCount,
	status,
	statusHappy,
	showDifficultyButtons,
	cardStats,
	intervalPreviews,
	onShowNext,
	onAddToQueue,
	onShowQueue,
	onMarkAgain,
	onMarkHard,
	onMarkGood,
	onMarkEasy,
}) => {
	return (
		<div>
			<h4>Incremental Reading</h4>

			{/* Queue stats */}
			<div
				style={{
					marginBottom: "15px",
					padding: "10px",
					backgroundColor: "var(--background-secondary)",
					borderRadius: "5px",
				}}
			>
				<div style={{ marginBottom: "5px" }}>Due: {dueCount}</div>
				<div>Total in queue: {totalCount}</div>
			</div>

			{/* Show Next button - primary only when there are notes to show */}
			<button
				className={dueCount > 0 ? "mod-cta" : ""}
				style={{ marginBottom: "10px", width: "100%" }}
				onClick={onShowNext}
			>
				Show Next
			</button>

			{/* Add to Queue button */}
			<button
				style={{ marginBottom: "10px", width: "100%" }}
				onClick={onAddToQueue}
			>
				Add Current Note to Queue
			</button>

			{/* Card Statistics (when reviewing) */}
			{cardStats && (
				<div
					style={{
						marginTop: "15px",
						padding: "10px",
						backgroundColor: "var(--background-secondary)",
						borderRadius: "5px",
						fontSize: "0.9em",
					}}
				>
					<div style={{ marginBottom: "3px" }}>
						<strong>Card Stats:</strong>
					</div>
					<div style={{ marginBottom: "2px" }}>
						Memory: {cardStats.stability}d | Difficulty:{" "}
						{cardStats.difficulty}/10
					</div>
					<div style={{ fontSize: "0.85em", opacity: 0.8 }}>
						Reviews: {cardStats.reps} | Forgotten:{" "}
						{cardStats.lapses}
					</div>
				</div>
			)}

			{/* Status message area */}
			<div
				style={{
					marginTop: "20px",
					padding: "10px",
					textAlign: "center",
					color: statusHappy
						? "var(--text-success)"
						: "var(--text-normal)",
				}}
			>
				{status}
				{showDifficultyButtons && intervalPreviews && (
					<div
						style={{
							marginTop: "10px",
							marginBottom: "5px",
							fontSize: "0.85em",
							opacity: 0.7,
						}}
					>
						Next review intervals:
					</div>
				)}
				{showDifficultyButtons && (
					<div
						style={{
							display: "grid",
							gridTemplateColumns: "1fr 1fr",
							gap: "8px",
							marginTop: "10px",
						}}
					>
						<button style={{ flex: "1" }} onClick={onMarkAgain}>
							Again
							{intervalPreviews && (
								<div
									style={{ fontSize: "0.75em", opacity: 0.7 }}
								>
									{intervalPreviews[Rating.Again]}
								</div>
							)}
						</button>
						<button style={{ flex: "1" }} onClick={onMarkHard}>
							Hard
							{intervalPreviews && (
								<div
									style={{ fontSize: "0.75em", opacity: 0.7 }}
								>
									{intervalPreviews[Rating.Hard]}
								</div>
							)}
						</button>
						<button
							className="mod-cta"
							style={{ flex: "1" }}
							onClick={onMarkGood}
						>
							Good
							{intervalPreviews && (
								<div
									style={{ fontSize: "0.75em", opacity: 0.7 }}
								>
									{intervalPreviews[Rating.Good]}
								</div>
							)}
						</button>
						<button style={{ flex: "1" }} onClick={onMarkEasy}>
							Easy
							{intervalPreviews && (
								<div
									style={{ fontSize: "0.75em", opacity: 0.7 }}
								>
									{intervalPreviews[Rating.Easy]}
								</div>
							)}
						</button>
					</div>
				)}
			</div>

			{/* Show Queue button - less prominent, at the bottom */}
			<button
				style={{
					marginTop: "20px",
					width: "100%",
					fontSize: "0.9em",
					opacity: "0.7",
				}}
				onClick={onShowQueue}
			>
				Show Queue (debug)
			</button>
		</div>
	);
};
