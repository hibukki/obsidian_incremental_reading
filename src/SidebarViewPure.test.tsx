import React from "react";
import { render } from "@testing-library/react";
import { Rating, State } from "ts-fsrs";
import { SidebarViewPure, SidebarViewPureProps } from "./SidebarViewPure";

// Mock handlers
const mockHandlers = {
	onShowNext: jest.fn(),
	onAddToQueue: jest.fn(),
	onShowQueue: jest.fn(),
	onMarkAgain: jest.fn(),
	onMarkHard: jest.fn(),
	onMarkGood: jest.fn(),
	onMarkEasy: jest.fn(),
};

describe("SidebarViewPure", () => {
	describe("Snapshot Tests", () => {
		test("renders correctly when there are due notes and user is reviewing one", () => {
			const props: SidebarViewPureProps = {
				dueNowCount: 5,
				dueTodayCount: 12,
				totalCount: 25,
				status: "How was this note? (Use commands or click below)",
				statusHappy: false,
				showDifficultyButtons: true,
				cardStats: {
					stability: 3.5,
					difficulty: 6.2,
					reps: 4,
					lapses: 1,
					state: State.Review,
				},
				intervalPreviews: {
					[Rating.Again]: "10m",
					[Rating.Hard]: "2h",
					[Rating.Good]: "1d",
					[Rating.Easy]: "3d",
				},
				...mockHandlers,
			};

			const { container } = render(<SidebarViewPure {...props} />);
			expect(container).toMatchSnapshot();
		});

		test("renders correctly when there are no notes to review", () => {
			const props: SidebarViewPureProps = {
				dueNowCount: 0,
				dueTodayCount: 0,
				totalCount: 15,
				status: "ðŸŽ‰ Done for today! All caught up!",
				statusHappy: true,
				showDifficultyButtons: false,
				cardStats: null,
				intervalPreviews: null,
				...mockHandlers,
			};

			const { container } = render(<SidebarViewPure {...props} />);
			expect(container).toMatchSnapshot();
		});
	});

	describe("Additional Scenarios", () => {
		test("renders correctly when there are notes due later today", () => {
			const props: SidebarViewPureProps = {
				dueNowCount: 0,
				dueTodayCount: 8,
				totalCount: 20,
				status: "",
				statusHappy: false,
				showDifficultyButtons: false,
				cardStats: null,
				intervalPreviews: null,
				...mockHandlers,
			};

			const { container } = render(<SidebarViewPure {...props} />);
			expect(container).toMatchSnapshot();
		});

		test("renders correctly with status message after rating", () => {
			const props: SidebarViewPureProps = {
				dueNowCount: 3,
				dueTodayCount: 7,
				totalCount: 18,
				status: "Scheduled for 2 hours",
				statusHappy: false,
				showDifficultyButtons: false,
				cardStats: null,
				intervalPreviews: null,
				...mockHandlers,
			};

			const { container } = render(<SidebarViewPure {...props} />);
			expect(container).toMatchSnapshot();
		});
	});
});
