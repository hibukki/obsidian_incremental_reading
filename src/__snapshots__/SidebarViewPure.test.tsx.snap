// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarViewPure Additional Scenarios renders correctly when current note is in queue (not reviewing) 1`] = `
<div>
  <div>
    <h4>
      Incremental Reading
    </h4>
    <button
      class="mod-cta"
      style="margin-bottom: 10px; width: 100%;"
    >
      Show next (5 left)
    </button>
    <div
      style="margin-bottom: 10px; padding: 10px; border-radius: 5px; text-align: center; opacity: 0.8;"
    >
      Note in queue (due in
       
      5 days
      )
    </div>
    <details
      style="margin-top: 20px; font-size: 0.85em; opacity: 0.7;"
    >
      <summary
        style="cursor: pointer; margin-bottom: 8px; font-size: 0.9em;"
      >
        Debug
      </summary>
      <div
        style="padding: 10px; border-radius: 5px;"
      >
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Current note:
          </strong>
           
          queued-note
        </div>
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Total in queue:
          </strong>
           
          20
        </div>
        <button
          style="width: 100%; font-size: 0.9em;"
        >
          Show Queue File
        </button>
      </div>
    </details>
  </div>
</div>
`;

exports[`SidebarViewPure Additional Scenarios renders correctly when current note is not in queue (not reviewing) 1`] = `
<div>
  <div>
    <h4>
      Incremental Reading
    </h4>
    <button
      class="mod-cta"
      style="margin-bottom: 10px; width: 100%;"
    >
      Show next (5 left)
    </button>
    <button
      style="margin-bottom: 10px; width: 100%;"
    >
      Add Current Note to Queue
    </button>
    <details
      style="margin-top: 20px; font-size: 0.85em; opacity: 0.7;"
    >
      <summary
        style="cursor: pointer; margin-bottom: 8px; font-size: 0.9em;"
      >
        Debug
      </summary>
      <div
        style="padding: 10px; border-radius: 5px;"
      >
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Current note:
          </strong>
           
          not-queued-note
        </div>
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Total in queue:
          </strong>
           
          20
        </div>
        <button
          style="width: 100%; font-size: 0.9em;"
        >
          Show Queue File
        </button>
      </div>
    </details>
  </div>
</div>
`;

exports[`SidebarViewPure Additional Scenarios renders correctly when there are notes due later today 1`] = `
<div>
  <div>
    <h4>
      Incremental Reading
    </h4>
    <button
      class=""
      style="margin-bottom: 10px; width: 100%;"
    >
      Show next (none due)
    </button>
    <button
      style="margin-bottom: 10px; width: 100%;"
    >
      Add Current Note to Queue
    </button>
    <details
      style="margin-top: 20px; font-size: 0.85em; opacity: 0.7;"
    >
      <summary
        style="cursor: pointer; margin-bottom: 8px; font-size: 0.9em;"
      >
        Debug
      </summary>
      <div
        style="padding: 10px; border-radius: 5px;"
      >
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Current note:
          </strong>
           
          (no active note)
        </div>
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Total in queue:
          </strong>
           
          20
        </div>
        <button
          style="width: 100%; font-size: 0.9em;"
        >
          Show Queue File
        </button>
      </div>
    </details>
  </div>
</div>
`;

exports[`SidebarViewPure Additional Scenarios renders correctly with status message after rating 1`] = `
<div>
  <div>
    <h4>
      Incremental Reading
    </h4>
    <button
      class="mod-cta"
      style="margin-bottom: 10px; width: 100%;"
    >
      Show next (3 left)
    </button>
    <button
      style="margin-bottom: 10px; width: 100%;"
    >
      Add Current Note to Queue
    </button>
    <div
      style="margin-top: 10px; padding: 10px; text-align: center;"
    >
      Scheduled for 2 hours
    </div>
    <details
      style="margin-top: 20px; font-size: 0.85em; opacity: 0.7;"
    >
      <summary
        style="cursor: pointer; margin-bottom: 8px; font-size: 0.9em;"
      >
        Debug
      </summary>
      <div
        style="padding: 10px; border-radius: 5px;"
      >
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Current note:
          </strong>
           
          test-note
        </div>
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Total in queue:
          </strong>
           
          18
        </div>
        <button
          style="width: 100%; font-size: 0.9em;"
        >
          Show Queue File
        </button>
      </div>
    </details>
  </div>
</div>
`;

exports[`SidebarViewPure Snapshot Tests renders correctly when there are due notes and user is reviewing one 1`] = `
<div>
  <div>
    <h4>
      Incremental Reading
    </h4>
    <button
      class="mod-cta"
      disabled=""
      style="margin-bottom: 10px; width: 100%;"
    >
      Show next (5 left)
    </button>
    <div
      style="margin-bottom: 10px; text-align: center; font-size: 0.9em; font-weight: bold;"
    >
      How was this note?
    </div>
    <div
      style="display: grid; grid-template-columns: 1fr 1fr; gap: 8px; margin-bottom: 15px;"
    >
      <button
        style="flex: 1;"
      >
        Again
        <div
          style="font-size: 0.75em; opacity: 0.7;"
        >
          10m
        </div>
      </button>
      <button
        style="flex: 1;"
      >
        Hard
        <div
          style="font-size: 0.75em; opacity: 0.7;"
        >
          2h
        </div>
      </button>
      <button
        class="mod-cta"
        style="flex: 1;"
      >
        Good
        <div
          style="font-size: 0.75em; opacity: 0.7;"
        >
          1d
        </div>
      </button>
      <button
        style="flex: 1;"
      >
        Easy
        <div
          style="font-size: 0.75em; opacity: 0.7;"
        >
          3d
        </div>
      </button>
    </div>
    <div
      style="margin-bottom: 15px; padding: 10px; border-radius: 5px;"
    >
      <div
        style="margin-bottom: 8px; font-weight: bold; font-size: 0.9em;"
      >
        Priority:
      </div>
      <div
        style="display: flex; gap: 6px;"
      >
        <button
          style="flex: 1; padding: 6px; border-radius: 4px; cursor: pointer; font-weight: normal;"
        >
          High
        </button>
        <button
          style="flex: 1; padding: 6px; border-radius: 4px; cursor: pointer; font-weight: bold;"
        >
          Normal
        </button>
        <button
          style="flex: 1; padding: 6px; border-radius: 4px; cursor: pointer; font-weight: normal;"
        >
          Low
        </button>
      </div>
    </div>
    <details
      style="margin-top: 20px; font-size: 0.85em; opacity: 0.7;"
    >
      <summary
        style="cursor: pointer; margin-bottom: 8px; font-size: 0.9em;"
      >
        Debug
      </summary>
      <div
        style="padding: 10px; border-radius: 5px;"
      >
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Current note:
          </strong>
           
          test-note
        </div>
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Total in queue:
          </strong>
           
          25
        </div>
        <button
          style="width: 100%; font-size: 0.9em;"
        >
          Show Queue File
        </button>
        <div
          style="margin-top: 10px; padding-top: 10px;"
        >
          <div
            style="margin-bottom: 4px;"
          >
            <strong>
              Card Stats:
            </strong>
          </div>
          <div
            style="margin-bottom: 2px;"
          >
            Memory: 
            3.5
            d | Difficulty:
             
            6.2
            /10
          </div>
          <div
            style="font-size: 0.9em; opacity: 0.8;"
          >
            Reviews: 
            4
             | Forgotten:
             
            1
          </div>
        </div>
      </div>
    </details>
  </div>
</div>
`;

exports[`SidebarViewPure Snapshot Tests renders correctly when there are no notes to review 1`] = `
<div>
  <div>
    <h4>
      Incremental Reading
    </h4>
    <button
      class=""
      style="margin-bottom: 10px; width: 100%;"
    >
      Show next (none due)
    </button>
    <button
      style="margin-bottom: 10px; width: 100%;"
    >
      Add Current Note to Queue
    </button>
    <div
      style="margin-top: 10px; padding: 10px; text-align: center;"
    >
      ðŸŽ‰ Done for today! All caught up!
    </div>
    <details
      style="margin-top: 20px; font-size: 0.85em; opacity: 0.7;"
    >
      <summary
        style="cursor: pointer; margin-bottom: 8px; font-size: 0.9em;"
      >
        Debug
      </summary>
      <div
        style="padding: 10px; border-radius: 5px;"
      >
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Current note:
          </strong>
           
          test-note
        </div>
        <div
          style="margin-bottom: 8px;"
        >
          <strong>
            Total in queue:
          </strong>
           
          15
        </div>
        <button
          style="width: 100%; font-size: 0.9em;"
        >
          Show Queue File
        </button>
      </div>
    </details>
  </div>
</div>
`;
